design-proposals/sriov-interfaces-mapping.md

```diff
@@ -0,0 +1,694 @@
+
+### Annotations Exposure Method
+A subset of the information in the networks-status annotation needs to be passed down to the virt-launcher for composing the right domain configuration.
+
+There have been several methods explored, recorded in [Appendix 1: Alternative Annotations Exposure Methods](sriov-interfaces-mapping.md#Appendix-1:-Alternative-Annotations-Exposure-Methods).
+
+In this section the chosen method is explained.
+
+#### Expose networks-status annotation to virt-launcher Pod
+The annotation content can be exposed inside virt-launcher Pod using [Kubernetes Downward API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/):
+
+virt-controller creates `virt-metadata` annotation that will hold generic information about the pod that we want to expose to the virt-launcher pod user.
+
+1. Given the networks-status annotation, process the content and compose 1:1 mapping between a (VMI) network from it and the device PCI address information and add it to virt-metadata annotation.
+For example:
+```yaml
+kubevirt.io/virt-metadata: |-
+{
+  network-to-pci-address: 
+  {
+    “sriovnet-vlan100-primary-mac”: “0000:04:02.5”,
+    “sriovnet-vlan200-primary-mac”: “0000:04:02.2”
+  }
+}
+```
+
+2. Mount `virt-metadata` annotation using Downward API as part of virt-launcher Pod creation:
+```yaml
+kind: VirtualMachineInstance
+metadata:
+  name: sriovvmi
+...
+spec:
+containers:
+- name: “compute”
+  ...
+  volumeMounts:
+    - name: virt-metadata-annotation
+      mountPath: /etc/podinfo
+      ...
+      volumes:
+- name: virt-metadata-annotation
+  downwardAPI:
+  items:
+    - path: "virt-metadata"
+      fieldRef:
+      fieldPath: metadata.annotations['virt-metadata']
+      ...
+```
+
+> **_Note_**: The annotation details will be populated at `/etc/podinfo/virt-metadata` file inside the pod.
+> Any change to the annotation will be reflected in the file.
+
+3. As part of virt-launcher VMI spec rendering, given the downward API file exists (e.g: `/etc/podinfo/virt-metadata` file),
+access the network-to-pci-address data, lookup for the PCI address based on the network name and create each SR-IOV HostDevice accordingly.
+
+4. If the file doesn't exist, fall back to the legacy mapping method (based on environment variables).
+
+##### Backward compatibility
+- On post Kubevirt upgrade new virt-controller will treat new and old VMIs and their pod in the same manner - adding the virt-metadata annotation on the pods.
+
+- No disruption to already existing VMIs workloads, though the SR-IOV mapping may not be correct.
+
+##### Pros
+1. This solution involves only virt-controller and virt-launcher.
+2. The mapping info is passed down using the standard Kubernetes mechanism (Downward API) with no need to add additional logic.
+3. Flexibility, no API change to VM or VMI objects.
+
+##### Cons
+2. Depends on Downward API feature and volume mount on the Pod.
```

Comment:
In general I just prefer to not add more mounts. I am pretty sure it is not a problem to mount to a few pods the volume where a sriov device is required, it may be a problem if we add it to all pods. If possible I would want to avoid it, but without further investigations I would have no prove that it would create an issue. If we can avoid it (for a big set of pods), we just don't have to investigate it.